README.md
<div align="center">
  <a href="https://github.com/webpack/webpack">
    <img width="200" height="200"
      src="https://webpack.js.org/assets/icon-square-big.svg">
  </a>
</div>

[![npm][npm]][npm-url]
[![node][node]][node-url]
[![install size](https://packagephobia.com/badge?p=generate-sitemap-webpack-plugin)](https://packagephobia.com/result?p=generate-sitemap-webpack-plugin)

# generate-sitemap-webpack-plugin

Webpack plugin to generate a sitemap.

## Getting Started

To begin, you'll need to install generate-sitemap-webpack-plugin:

```shell
npm install generate-sitemap-webpack-plugin --save-dev
```

```shell
yarn add -D generate-sitemap-webpack-plugin
```

```shell
pnpm add -D generate-sitemap-webpack-plugin
```

Then add the plugin to your webpack config. For example:

**webpack.config.js**

```javascript
const SitemapPlugin = require('generate-sitemap-webpack-plugin');

module.exports = {
  plugins: [new SitemapPlugin({ baseURL: 'https://your.website' })],
};
```

**sitemap.xml**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://your.website/index.html</loc>
  </url>
</urlset>
```

> **Note**
>
> `generate-sitemap-webpack-plugin` is to use files that already exist in the source tree, as part of the build process.

> **Note**
>
> If you want to add custom locations for sitemap.xml, you can use urls options

<br/>
<br/>

## Usage

The plugin's signature:

**webpack.config.js**

```javascript
const SitemapPlugin = require('generate-sitemap-webpack-plugin');

module.exports = {
  plugins: [new SitemapPlugin({ baseURL, urls, emitted, chunk, options })],
};
```

- **[`baseURL`](#baseURL)**
- **[`urls`](#urls)**
- **[`emitted`](#emitted)**
- **[`chunk`](#chunk)**
- **[`options`](#options)**

<br/>

### `baseURL` <small>(required)</small>

Root URL of your site (e.g. https://your.website)

Type: `string`

<br/>

### `urls`

Optional array of locations on your site. These can be strings or you can provide object to customize each url.

Type:

```typescript
type urls = Array<string | SitemapURL>;

type SitemapURL = {
  loc: string;
  lastmod?: string;
  priority?: number;
  changefreq?: Changefreq;
};
```

> **Note**
>
> If you provide the object in array, the following attributes may be set on each path object.

| Name                 | Type     | Default     | Description                                                                                                                                                                                                     |
| -------------------- | -------- | ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `loc`<br/>(required) | `string` | N/A         | URL of the page (e.g. some/link). Sum of this value and baseURL must be less than 2,048 characters.                                                                                                             |
| `lastmod`            | `string` | `undefined` | The date of last modification of the page (e.g. 2023-12-08). This date should be in W3C Datetime format. This format allows you to omit the time portion, if desired, and use YYYY-MM-DD.                       |
| `priority`           | `number` | `undefined` | The priority of this URL relative to other URLs on your site. Valid values range from 0.0 to 1.0.                                                                                                               |
| `changefreq`         | `string` | `undefined` | How frequently the page is likely to change. List of applicable values based on [sitemaps.org](https://www.sitemaps.org/protocol.html):<br/>`always`, `hourly`, `daily`, `weekly`, `monthly`, `yearly`, `never` |

<br/>

### `emitted`

Optional object to customize each url by webpack emitted assets. If you set to boolean `true`, the emitted all `.html` file being used

- **[`callback`](#callback)**
- **[`pattern`](#pattern)**

#### `callback`

Callback function for use emitted asset filename

Type:

```typescript
type callback = (
  location: string
) => Omit<SitemapURL, 'loc'> | undefined | null;
```

<small>

- **[sitemapURL](#urls)**

</small>

Default: `N/A`

#### `pattern`

Specific pattern to filter the asset (e.g. .html), This can be string (glob pattern) or you can provide function instead of string pattern

Type:

```typescript
type pattern = string | ((asset: string) => boolean) | undefined;
```

Default: `**/*.html`

> Note
>
> `pattern` use glob pattern to filter asset.
> For more information, follow the links below
>
> - [Glob Pattern Reference | Visual Studio Code](https://code.visualstudio.com/docs/editor/glob-patterns)
> - [Featrues | minimatch](https://github.com/isaacs/minimatch#features)

<br/>

### `chunk`

Optional object to customize each url by webpack chunk. You can use auxiliary files to make sitemap include image

- **[`callback`](#callback)**
- **[`img`](#pattern)**

#### `callback`

Callback function for use chunk

Type <small>(Required)</small> :

```typescript
type callback = (
  name: string,
  hash: string
) => SitemapURL | string | undefined | null;
```

Default: `N/A`

#### `img`

Options for add image locations

Type: `boolean`

Default: `true`

> Note
>
> If you provide `options.chunk.img = true` each url automatically contains auxiliary files which image extension is matched
> `imageExtList = [apng, avif, gif, jpg, jpeg, jfif, pjpeg, pjp, png, svg, webp, bmp, ico, cur, tif, tiff]`

<br/>

### `options`

Optional object of configuration settings.

| Name         | Type                 | Default       | Description                                                                                                                                         |
| ------------ | -------------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| `filename`   | `string`             | `sitemap.xml` | Name of the sitemap file emitted to your build output                                                                                               |
| `format`     | `boolean`            | `false`       | Settings for format sitemap file. You can provide formatting options for write file prettier                                                        |
| `gzip`       | `boolean`            | `true`        | Generating a gzipped `.xml.gz` sitemap. You can provide false to skip generating a gzipped. By default, both xml files are generated                |
| `lastmod`    | `string` / `boolean` | `false`       | The date for <lastmod> on all urls. Can be overridden by url-specific lastmod config. If value is true, the current date will be used for all urls. |
| `priority`   | `number`             | `undefined`   | A <priority> to be set globally on all locations. Can be overridden by url-specific priorty config.0.                                               |
| `changefreq` | `string`             | `undefined`   | A <changefreq> to be set globally on all locations. Can be overridden by url-specific changefreq config.                                            |

[npm]: https://img.shields.io/npm/v/generate-sitemap-webpack-plugin.svg
[npm-url]: https://www.npmjs.com/package/generate-sitemap-webpack-plugin
[node]: https://img.shields.io/node/v/generate-sitemap-webpack-plugin.svg
[node-url]: https://nodejs.org
[size]: https://packagephobia.now.sh/badge?p=generate-sitemap-webpack-plugin
[size-url]: https://packagephobia.now.sh/result?p=generate-sitemap-webpack-plugin

## Examples

### 1. Complex Example

**webpack.config.js**

```javascript
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const SitemapPlugin = require('generate-sitemap-webpack-plugin');

const entries = ['index', 'help', 'ejs'];
const entry = {};
const views = [];
entries.forEach((name) => {
  entry[name] = path.resolve(__dirname, `src/js/${name}.js`);

  const htmlPlugin = new HtmlWebpackPlugin({
    chunks: [name],
    minify: false,
    filename: `${name}.${name.includes('ejs') ? 'ejs' : 'html'}`,
    template: path.resolve(__dirname, `src/views/${name}.ejs`),
  });

  views.push(htmlPlugin);
});

module.exports = {
  mode: 'production',
  entry: {
    index: path.resolve(__dirname, 'path/to/index.js')
    entry: path.resolve(__dirname, 'path/to/entry.js')
    ejs: path.resolve(__dirname, 'path/to/ejs.js')
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'js/[name].js',
  },
  plugins: [
    new SitemapPlugin({
      baseURL: 'https://your.website',
      emitted: {
        callback: (location) => ({
          priority: location.includes('index') ? 1 : undefined,
          changefreq: 'yearly',
        }),
        pattern: `**/*.{html,ejs}`,
      },
      urls: [
        'first',
        { loc: 'second.html', lastmod: '2023-12-19', priority: 0.3 },
      ],
      options: {
        filename: 'my-sitemap.xml',
        format: true, // default: false
        gzip: true // default: true
        changefreq: 'monthly',
        lastmod: true,
        priority: 0.6,
      },
    }),
    new HtmlWebpackPlugin({
      chunks: ['index'],
      filename: 'index.html',
      template: path.resolve(__dirname, 'path/to/index.html'),
    }),
    new HtmlWebpackPlugin({
      chunks: ['entry'],
      filename: 'entry.html',
      template: path.resolve(__dirname, 'path/to/entry.html'),
    }),
    new HtmlWebpackPlugin({
      chunks: ['ejs'],
      filename: 'ejs.ejs',
      template: path.resolve(__dirname, 'path/to/ejs.html'),
    }),
  ],
};
```

**my-sitemap.xml**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://your.website/index.html</loc>
    <lastmod>2023-12-19T03:29:51.291Z</lastmod>
    <changefreq>yearly</changefreq>
    <priority>1</priority>
  </url>
  <url>
    <loc>https://your.website/first</loc>
    <lastmod>2023-12-19T03:29:51.291Z</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.6</priority>
  </url>
  <url>
    <loc>https://your.website/entry.html</loc>
    <lastmod>2023-12-19T03:29:51.291Z</lastmod>
    <changefreq>yearly</changefreq>
  </url>
  <url>
    <loc>https://your.website/ejs.ejs</loc>
    <lastmod>2023-12-19T03:29:51.291Z</lastmod>
    <changefreq>yearly</changefreq>
  </url>
  <url>
    <loc>https://your.website/second.html</loc>
    <lastmod>2023-12-19</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.3</priority>
  </url>
</urlset>
```

### 2. Split Example

**webpack.config.js**

```javascript
const SitemapPlugin = require('generate-sitemap-webpack-plugin');

module.exports = {
  // Some webpack config
  ...,
  plugins: [
    new SitemapPlugin({
      baseURL: 'https://your.website',
      urls: Array.from({ length: 150000 }, (_, i) => ({
        loc: `${i}.html`,
        lastmod:
          i < 50000 ? `2023-12-19` : undefined,
      })),
      options: {
        format: true
      }
    }),
  ],
};
```

**sitemap.xml**

```xml
<?xml version="1" encoding="UTF-8"?>
<sitemapindex>
  <sitemap>
    <loc>https://your.website/sitemap1.xml</loc>
    <lastmod>2023-12-19</lastmod>
  </sitemap>
  <sitemap>
    <loc>https://your.website/sitemap2.xml</loc>
  </sitemap>
  <sitemap>
    <loc>https://your.website/sitemap3.xml</loc>
  </sitemap>
</sitemapindex>
```

**sitemap1.xml**

```xml
<?xml version="1" encoding="UTF-8"?>
<urlset>
  <url>
    <loc>https://your.website/0.html</loc>
    <lastmod>2023-12-19</lastmod>
  </url>
  <url>
    <loc>https://your.website/1.html</loc>
    <lastmod>2023-12-19</lastmod>
  </url>
  ... 49,998 items
</urlset>
```

### 3. Chunk Example

**webpack.config.js**

```javascript
const SitemapPlugin = require('generate-sitemap-webpack-plugin');

module.exports = {
  // Some webpack config
  ...,
  entry: {
    index: 'path/to/index.js'
    skip: 'path/to/skip.js'
  },
  module: {
    rules: [
      {
        test: /\.(?:jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$/i,
        type: 'asset/resource',
        generator: {
          filename: 'assets/[name]-[hash][ext]',
        },
      },
    ],
  },
  plugins: [
    new SitemapPlugin({
      baseURL: 'https://your.website',
      chunk: {
        callback: (name) => (name === 'skip' ? null : name + '.html'),
      },
      emitted: false,
      options: {
        format: true,
      },
    }),
  ],
};
```

**index.js**

```javascript
import Sample from 'path/to/asset/sample.png';
import Move from 'path/to/asset/move.gif';

... some scripts
```

**sitemap.xml**

```xml
<?xml version="1" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">
  <url>
    <loc>https://your.website/index.html</loc>
    <image:image>
      <image:loc>https://your.website/assets/move-31d6cfe0d16ae931b73c.gif</image:loc>
    </image:image>
    <image:image>
      <image:loc>https://your.website/assets/sample-31d6cfe0d16ae931b73c.png</image:loc>
    </image:image>
  </url>
</urlset>

```

## License

[MIT](./LICENSE)

[npm]: https://img.shields.io/npm/v/generate-sitemap-webpack-plugin.svg
[npm-url]: https://npmjs.com/package/generate-sitemap-webpack-plugin
[node]: https://img.shields.io/node/v/generate-sitemap-webpack-plugin.svg
[node-url]: https://nodejs.org


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v1.0.1...v1.0.2) (2023-12-28)

- remove not used dependencies
- username update

### [1.0.1](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v1.0.0...v1.0.1) (2023-12-20)

### Bug Fixes

- add chunk url validation ([c801bce](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/c801bce9a7105897041a8d900f9523b97b5ec81a))

## [1.0.0](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.3.0...v1.0.0) (2023-12-20)

### Features

- options.chunk ([e415b75](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/e415b759fa609c532708ae87c12870661300402b))

### Bug Fixes

- resolve errors that do not have attributes attached ([8f76083](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/8f76083cc1c64420ee9575825188d906fa65c80b))

## [0.3.0](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.2.0...v0.3.0) (2023-12-19)

### Features

- gzip options ([092d0b8](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/092d0b8fbdb14cfaa382dd18ec7b0cf44b459ee3))

## [0.2.0](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.1.2...v0.2.0) (2023-12-19)

### ⚠ BREAKING CHANGES

- sitemap plugin automaticaly split sitemap file when sitemap has more than 50,000 URLs
- you can provide formatting options for write file prettier

### Features

- add split sitemap, format options ([dea4b65](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/dea4b65b6d1ae2f264728b9892f7f961577f7070))

### [0.1.2](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.1.1...v0.1.2) (2023-12-18)

### Features

- change xml creating method ([273b593](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/273b5934b18bc9b713504b575da6f3e51f5666a0))

### Bug Fixes

- change copy option method (Object.assing => spread) ([2112613](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/21126138c45f6ecfc3c0dbb4ad9ddfac4db624d0))
- options validation ([782f360](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/782f360e2eebbbb6b8d80bfb4567eb5bc5734ad6))

### [0.1.1](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.1.0...v0.1.1) (2023-12-13)

### Bug Fixes

- options.baseURL validation error ([24a4634](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/24a46341b2c5a34657622c12a120b1148bccd840))

## [0.1.0](https://github.com/janghye0k/generate-sitemap-webpack-plugin/compare/v0.0.0...v0.1.0) (2023-12-12)

### ⚠ BREAKING CHANGES

- add emitted.pattern to filter asset (glob pattern)

### Features

- emitted.pattern ([d4966d1](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/d4966d17784b873ccb4e41815f356cca4210f1bd))

### Bug Fixes

- error when plugin option is undefined ([f6657a2](https://github.com/janghye0k/generate-sitemap-webpack-plugin/commit/f6657a2668a80b331ea2442131945957eae5fa22))


src/index.js
const path = require('path');
const zlib = require('zlib');
const fs = require('fs');
const schema = require('./options.json');
const { validate } = require('schema-utils');
const { minimatch } = require('minimatch');
const { XMLBuilder } = require('fast-xml-parser');
const { WebpackError } = require('webpack');

/** @typedef {import("schema-utils/declarations/validate").Schema} Schema */
/** @typedef {import("webpack").Compiler} Compiler */
/** @typedef {import("webpack").Compilation} Compilation */

/**
 * @typedef {'always' | 'hourly' | 'daily' | 'weekly' | 'monthly' | 'yearly' |'never'} Changefreq
 */

/**
 * @typedef {Object} SitemapURL
 * @property {string} loc URL of the page (e.g. some/link). Sum of this value and baseURL must be less than 2,048 characters.
 * @property {string} [lastmod] The date of last modification of the page (e.g. 2023-12-08). This date should be in W3C Datetime format. This format allows you to omit the time portion, if desired, and use YYYY-MM-DD.
 * @property {number} [priority] The priority of this URL relative to other URLs on your site. Valid values range from 0.0 to 1.0.
 * @property {Changefreq} [changefreq] How frequently the page is likely to change.
 */

/**
 * @typedef {Omit<SitemapURL, 'loc'> | null | undefined} EmittedCallbackReturns
 */

/**
 * @callback EmittedCallback
 * @param {string} location URL location created by asset filename (e.g. http://your.site/emitted.html)
 * @returns {EmittedCallbackReturns}
 */

/**
 * @typedef {Object} EmittedOptions
 * @property {EmittedCallback} callback
 * @property {string | ((asset: string) => boolean)} [pattern] Specific pattern to filter the asset (e.g. .html), This can be string (glob pattern) or you can provide function instead of string pattern
 */

/**
 * @callback ChunkCallback
 * @param {string} name
 * @param {string} hash
 * @returns {SitemapURL | string | undefined | null}
 */

/**
 * @typedef {Object} ChunkOptions
 * @property {ChunkCallback} callback
 * @property {boolean} img Options for add image sitemap (Default: true)
 */

/**
 * @typedef {Object} AdditionalOptions
 * @property {string} [filename] Name of the sitemap file emitted to your build output
 * @property {boolean} [format] Settings for format sitemap file
 * @property {boolean} [gzip] Generating a gzipped `.xml.gz` sitemap.  You can provide false to skip generating a gzipped. By default, both xml files are generated
 * @property {string | boolean} [lastmod] The date for <lastmod> on all urls. Can be overridden by url-specific lastmod config. If value is true, the current date will be used for all urls.
 * @property {number} [priority] A <priority> to be set globally on all locations. Can be overridden by url-specific priorty config.
 * @property {Changefreq} [changefreq] A <changefreq> to be set globally on all locations. Can be overridden by url-specific changefreq config.
 */

/**
 * @typedef {Object} PluginOptions
 * @property {string} baseURL Root URL of your site (e.g. https://your.site)
 * @property {Array<string | SitemapURL>} [urls] Optional array of locations on your site. These can be strings or you can provide object to customize each url.
 * @property {EmittedOptions | boolean} [emitted] Optional object to customize each url by webpack emitted assets
 * @property {ChunkOptions} [chunk] Optional object to customize each url by webpack chunk. You can use auxiliary file to make sitemap include image
 * @property {AdditionalOptions} [options] Optional object of configuration settings.
 */

const DEFAULT_PATTERN = '**/*.html';
const DEFAULT_NAME = 'sitemap.xml';
const DEFAULT_FORMAT = false;
const DEFAULT_GZIP = true;
const IMAGE_PATTERN =
  '**/*.{apng,avif,gif,jpg,jpeg,jfif,pjpeg,pjp,png,svg,webp,bmp,ico,cur,tif,tiff}';

const PLUGIN = 'GenerateSitemapWebpackPlugin';
/** @type {Record<string, { validate: (value: any) => boolean; schema: object }>} */
const URL_VALIDATION_MAP = {
  lastmod: {
    validate: (value) => typeof value === 'string',
    schema: schema.definitions.SitemapURL.properties.lastmod,
  },
  priority: {
    validate: (value) => 0 <= value && value <= 1,
    schema: schema.definitions.SitemapURL.properties.priority,
  },
  changefreq: {
    validate: (value) =>
      [
        'always',
        'hourly',
        'daily',
        'weekly',
        'monthly',
        'yearly',
        'never',
      ].includes(value),
    schema: schema.definitions.SitemapURL.properties.changefreq,
  },
};

const DEFAULT_CONFIG = {
  emitted: true,
};

const MAX_URL_SIZE = 50000;
const XML_PREFIX = `<?xml version="1" encoding="UTF-8"?>`;
const XML_ATTRS = {
  base: { '@_xmlns': 'http://www.sitemaps.org/schemas/sitemap/0.9' },
  image: { '@_xmlns:image': 'http://www.google.com/schemas/sitemap-image/1.1' },
};

class SitemapPlugin {
  /**
   * @private
   * @type {PluginOptions & { emitted: boolean | EmittedOptions }}
   */
  options;
  /**
   * @private
   * @type {Omit<SitemapURL, 'loc'>}
   */
  commonURLOptions;

  /**
   *
   * @param {PluginOptions} options
   */
  constructor(options) {
    const opts = { ...DEFAULT_CONFIG, ...(options || {}) };
    validate(/** @type {Schema} */ (schema), opts, {
      name: 'Sitemap Plugin',
      baseDataPath: 'options',
    });
    this.options = opts;

    // Set baseURL
    const { baseURL } = this.options;
    this.options.baseURL = baseURL.endsWith('/') ? baseURL : `${baseURL}/`;

    // Set common options
    const { lastmod, changefreq, priority } = this.options.options || {};
    this.commonURLOptions = { lastmod: undefined, changefreq, priority };

    if (lastmod === true) {
      this.commonURLOptions.lastmod = new Date().toISOString();
    } else if (typeof lastmod === 'string')
      this.commonURLOptions.lastmod = lastmod;
  }

  /**
   * @private
   * @template T
   * @param {Array<T>} arr
   * @param {number} size
   * @returns {Array<T[]>}
   */
  chunkArray(arr, size) {
    return Array.from({ length: Math.ceil(arr.length / size) }, (v, i) =>
      arr.slice(i * size, i * size + size)
    );
  }

  /**
   * @private
   * @param {SitemapURL[]} arr
   */
  maxLastmod(arr) {
    return arr.reduce(
      /** @param {string | undefined} result */
      (result, url) => {
        if (!url.lastmod) return result;
        if (!result) return url.lastmod;
        const a = new Date(result);
        const b = new Date(url.lastmod);
        return a > b ? result : url.lastmod;
      },
      undefined
    );
  }

  /**
   * @private
   * @param {SitemapURL[]} sitemapURLs
   */
  createSitemapData(sitemapURLs) {
    const imageURL = sitemapURLs.some(
      /** @param {any} url */
      (url) => Array.isArray(url['image:image'])
    );
    const data = {
      urlset: {
        ...XML_ATTRS.base,
        ...(imageURL ? XML_ATTRS.image : {}),
        url: sitemapURLs,
      },
    };
    return data;
  }

  /**
   * @private
   * @param {SitemapURL[]} sitemapURLs
   * @param {string} outputPath
   */
  createXMLSitemap(sitemapURLs, outputPath) {
    const filename = this.options.options?.filename || DEFAULT_NAME;

    const ext = path.extname(filename);
    const basename = filename.substring(0, filename.length - ext.length);
    const rootFilePath = path.join(outputPath, filename);
    const isFormat = this.options.options?.format ?? DEFAULT_FORMAT;
    const isGzip = this.options.options?.gzip ?? DEFAULT_GZIP;

    const builder = new XMLBuilder({
      format: isFormat,
      ignoreAttributes: false,
    });
    const chunkURLs = this.chunkArray(sitemapURLs, MAX_URL_SIZE);
    const chunkSize = chunkURLs.length;

    /** @type {Map<string, object>} */
    const xmlMap = new Map();
    if (chunkSize > 1) {
      // chunk case
      // create sitemapindex file
      const baseURL = this.options.baseURL;

      const xmlRoot = {
        ...XML_ATTRS.base,
        sitemapindex: {
          sitemap: chunkURLs.map((chunkURL, index) => {
            // create chunked sitemap
            const chunkFilename = `${basename}${index + 1}${ext}`;
            xmlMap.set(
              path.join(outputPath, chunkFilename),
              this.createSitemapData(chunkURL)
            );

            // create sitemap instance
            const sitemap = {
              loc: baseURL + chunkFilename,
              lastmod: this.maxLastmod(chunkURL),
            };
            if (isGzip) sitemap.loc += '.gz';
            return sitemap;
          }),
        },
      };
      xmlMap.set(rootFilePath, xmlRoot);
    } else xmlMap.set(rootFilePath, this.createSitemapData(sitemapURLs));

    xmlMap.forEach(async (xml, file) => {
      const data = XML_PREFIX + (isFormat ? `\n` : '') + builder.build(xml);
      fs.writeFileSync(file, data, 'utf-8');

      if (!isGzip) return;
      const compressed = zlib.gzipSync(Buffer.from(data));
      fs.writeFileSync(file + '.gz', compressed);
    });
  }

  /**
   * @private
   * @param {Compilation} compilation
   * @param {SitemapURL} sitemapURL
   */
  validateURL(compilation, sitemapURL) {
    Object.entries(sitemapURL).forEach(([k, v]) => {
      const item = URL_VALIDATION_MAP[k];
      if (!item || v === undefined || item.validate(v)) return;
      const error = new WebpackError(
        `Invalid options.emitted.callback\nthis callback's returnValue.${k} should be follow below description\n\n${Object.entries(
          item.schema
        )
          .map(([key, desc]) => `${key}: ${desc}`)
          .join('\n')}`
      );
      compilation.errors.push(error);
    });
    return sitemapURL;
  }

  /**
   * Generate sitemap data ues emittedAssets
   * @private
   * @param {Compilation} compilation
   * @returns {SitemapURL[]}
   */
  generateSitemapDataFromEmittedAssets(compilation) {
    const { baseURL, emitted } = this.options;

    /** @type {SitemapURL[]} */
    const sitemapURLs = [];
    if (emitted === false) return sitemapURLs;

    // add emitted url
    const emittedParam = emitted === true ? /**@type {any} */ ({}) : emitted;
    const callback = emittedParam?.callback || (() => ({}));
    const pattern = emittedParam?.pattern || DEFAULT_PATTERN;

    compilation.emittedAssets.forEach((assetName) => {
      if (typeof pattern === 'string') {
        if (!minimatch(assetName, pattern, { matchBase: true })) return;
      } else if (!pattern(assetName)) return;

      const location = baseURL + assetName;
      const result = callback(location) || {};
      if (!result || typeof result !== 'object')
        compilation.errors.push(
          new WebpackError(
            `options.emitted.callback should be return object. not to be ${typeof result}`
          )
        );

      /** @type {SitemapURL} */
      const { loc, lastmod, changefreq, priority } = {
        loc: location,
        ...this.commonURLOptions,
        ...result,
      };
      const sitemapURL = { loc, lastmod, changefreq, priority };
      this.validateURL(compilation, sitemapURL);
      sitemapURLs.push(sitemapURL);
    });

    return sitemapURLs;
  }

  /**
   * Generate sitemap data by custom url options
   * @private
   * @returns {SitemapURL[]}
   */
  generateSitemapDataFromURL() {
    const URLs = this.options.urls || [];

    return URLs.map((url) => {
      const { loc: location, ...item } =
        typeof url === 'string' ? { loc: url } : url;
      const link = location.startsWith('/') ? location.slice(1) : location;
      const { loc, lastmod, changefreq, priority } = {
        loc: this.options.baseURL + link,
        ...this.commonURLOptions,
        ...item,
      };
      return { loc, lastmod, changefreq, priority };
    });
  }

  /**
   * @private
   * @param {Compilation} compilation
   * @returns {SitemapURL[]}
   */
  generateSitemapDataFromChunks(compilation) {
    const { callback, img = true } = this.options.chunk || {};
    const { baseURL } = this.options;

    /** @type {SitemapURL[]} */
    const sitemapURLs = [];
    if (typeof callback !== 'function') return sitemapURLs;

    compilation.chunks.forEach((chunk) => {
      const { name, hash, auxiliaryFiles } = chunk;
      if (!name || !hash) return;
      const returnValues = callback(name, hash);
      if (!returnValues) return;

      const { loc, lastmod, changefreq, priority } = /** @type {SitemapURL} */ (
        typeof returnValues === 'string' ? { loc: returnValues } : returnValues
      );
      const location = loc.startsWith('/') ? loc.substring(1) : loc;
      /** @type {any} */
      const sitemapURL = {
        loc: baseURL + location,
        lastmod,
        changefreq,
        priority,
      };
      this.validateURL(compilation, sitemapURL);
      sitemapURLs.push(sitemapURL);

      if (!img) return;
      /** @type {string[]} */
      const images = [];
      auxiliaryFiles.forEach((file) => {
        if (minimatch(file, IMAGE_PATTERN, { matchBase: true }))
          images.push(file);
      });
      if (images.length === 0) return;

      sitemapURL['image:image'] = images.map((img) => ({
        'image:loc': baseURL + img,
      }));
    });

    return sitemapURLs;
  }

  /** @param {Compiler} compiler */
  apply(compiler) {
    compiler.hooks.afterEmit.tapAsync(PLUGIN, (compilation, done) => {
      if (compilation.options.mode !== 'production') return done();

      const sitemapURLs = [
        ...this.generateSitemapDataFromEmittedAssets(compilation),
        ...this.generateSitemapDataFromURL(),
        ...this.generateSitemapDataFromChunks(compilation),
      ].sort((a, b) => {
        const comapre = (b.priority ?? 0.5) - (a.priority ?? 0.5);
        return comapre;
      });

      // create sitemap
      const outputPath = compilation.options.output.path;
      this.createXMLSitemap(sitemapURLs, /** @type {any} */ (outputPath));

      done();
    });
  }
}

module.exports = SitemapPlugin;
